File: src/main/java/recipeapplication/security/WebSecurityConfig.java
package recipeapplication.security;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.authentication.dao.DaoAuthenticationProvider;
import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.builders.WebSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.crypto.password.PasswordEncoder;
import recipeapplication.service.RecipeUserDetailsService;

@Configuration
@EnableWebSecurity
public class WebSecurityConfig extends WebSecurityConfigurerAdapter
{
    private RecipeUserDetailsService userDetailsService;

    @Autowired
    public WebSecurityConfig(RecipeUserDetailsService userDetailsService)
    {
        this.userDetailsService = userDetailsService;
    }

    @Override
    protected void configure(HttpSecurity http) throws Exception
    {
        // TODO auto login failing after change password

        // TODO enable the CSRF token
        http.csrf().disable()
            .authorizeRequests()
                .antMatchers("/signup", "/resetPassword", "/changePassword", "/api/signup", "/css/**", "/images/**", "/js/**")
                .permitAll()
                .antMatchers("/", "/recipe", "/deleteAccount").hasAuthority(Role.USER.toString())
                .antMatchers("/api/change_password").hasAuthority(Role.CHANGE_PASSWORD.toString())
                .and()
            .formLogin()
                .loginPage("/login")
                .permitAll()
                .and()
            .exceptionHandling()
                .accessDeniedPage("/login")
                .and()
            .logout().logoutSuccessUrl("/login").deleteCookies("JSESSIONID").invalidateHttpSession(true);
    }

    @Override
    public void configure(WebSecurity web)
    {
        // TODO there might be a better way to do this
        web.ignoring().antMatchers("/api/signup", "/api/password_reset");
    }

    @Override
    protected void configure(AuthenticationManagerBuilder auth)
    {
        auth.authenticationProvider(authenticationProvider());
    }

    @Bean
    public DaoAuthenticationProvider authenticationProvider()
    {
        DaoAuthenticationProvider authProvider = new DaoAuthenticationProvider();

        authProvider.setUserDetailsService(userDetailsService);
        authProvider.setPasswordEncoder(encoder());

        return authProvider;
    }

    @Bean
    public PasswordEncoder encoder()
    {
        return new BCryptPasswordEncoder(11);
    }
}

File: src/main/kotlin/thecookingpot/config/WebSecurityConfig.kt
package thecookingpot.config

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.config.Customizer.withDefaults
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.security.oauth2.client.oidc.userinfo.OidcUserRequest
import org.springframework.security.oauth2.client.oidc.userinfo.OidcUserService
import org.springframework.security.oauth2.client.userinfo.OAuth2UserService
import org.springframework.security.oauth2.core.oidc.user.OidcUser
import thecookingpot.exception.UserNotFoundException
import thecookingpot.security.AuthenticationPrincipal
import thecookingpot.service.UserService


@Configuration
@EnableWebSecurity
/**
 * TODO
 * - logout
 * - exception handling
 * - redundant code
 * - change password
 */
class WebSecurityConfig @Autowired constructor(private val userService: UserService) : WebSecurityConfigurerAdapter() {
    override fun configure(http: HttpSecurity) {
        /**
         * should be http {...}
         */
        http.csrf().disable()
            .authorizeRequests()
            .antMatchers(
                "/api/password-reset",
                "/api/signup",
                "/signup",
                "/reset-password",
                "/change-password-with-token",
                "/api/signup",
                "/css/**",
                "/images/**",
                "/js/**",
                "/react/**",
            ).permitAll()
            .antMatchers(
                "/",
                "/recipe",
                "/shopping-list",
                "/inventory",
                "/delete-account",
                "/change-password",
            ).authenticated()
            .and()
            .oauth2Login(withDefaults())
//            .exceptionHandling()
//                .accessDeniedPage("/login")
//                .and()
//            .logout().logoutSuccessUrl("/login").deleteCookies("JSESSIONID").invalidateHttpSession(true)
    }

    @Bean // TODO not required anymore
    open fun encoder(): PasswordEncoder {
        return BCryptPasswordEncoder(11)
    }

    @Bean
    fun oidcUserService(): OAuth2UserService<OidcUserRequest, OidcUser> {
        val delegate = OidcUserService()

        return OAuth2UserService { userRequest ->
            val oidcUser = delegate.loadUser(userRequest)
            val username = oidcUser.subject

            oidcUser.accessTokenHash

            val user = try {
                userService.getUser(username)
            } catch (e: UserNotFoundException) {
                userService.createUser(username, oidcUser.getClaim("email"))
            }

            AuthenticationPrincipal(user, oidcUser)
        }
    }
}